internal/xds/translator/ratelimit.go

```diff
@@ -329,70 +410,153 @@ func GetRateLimitServiceConfigStr(pbCfg *rlsconfv3.RateLimitConfig) (string, err
 	return buf.String(), err
 }
 
-// BuildRateLimitServiceConfig builds the rate limit service configuration based on
+// BuildRateLimitServiceConfig builds the rate limit service configurations based on
 // https://github.com/envoyproxy/ratelimit#the-configuration-format
-func BuildRateLimitServiceConfig(irListener *ir.HTTPListener) *rlsconfv3.RateLimitConfig {
-	pbDescriptors := make([]*rlsconfv3.RateLimitDescriptor, 0, len(irListener.Routes))
+// It returns a list of unique configurations, one for each domain needed across all listeners.
+func BuildRateLimitServiceConfig(irListeners []*ir.HTTPListener) []*rlsconfv3.RateLimitConfig {
+	// Map to store descriptors for each domain
+	domainDescriptors := make(map[string][]*rlsconfv3.RateLimitDescriptor)
+
+	// Process each listener
+	for _, irListener := range irListeners {
+		// Process each route to build descriptors
+		for _, route := range irListener.Routes {
+			if !routeContainsGlobalRateLimit(route) {
+				continue
+			}
 
-	for _, route := range irListener.Routes {
-		if routeContainsGlobalRateLimit(route) {
-			serviceDescriptors := buildRateLimitServiceDescriptors(route.Traffic.RateLimit.Global)
-
-			// Get route rule descriptors within each route.
-			//
-			// An example of route descriptor looks like this:
-			//
-			// descriptors:
-			//   - key:   ${RouteDescriptor}
-			//     value: ${RouteDescriptor}
-			//     descriptors:
-			//       - key:   ${RouteRuleDescriptor}
-			//         value: ${RouteRuleDescriptor}
-			//       - ...
-			//
-			routeDescriptor := &rlsconfv3.RateLimitDescriptor{
-				Key:         getRouteDescriptor(route.Name),
-				Value:       getRouteDescriptor(route.Name),
-				Descriptors: serviceDescriptors,
+			// Get route rule descriptors within each route
+			serviceDescriptors := buildRateLimitServiceDescriptors(route)
+			if len(serviceDescriptors) == 0 {
+				continue
+			}
+
+			// Determine the domain for this route
+			domain := irListener.Name // Default domain
+			if isSharedRateLimit(route) {
+				domain = getDomainName(route)
+			}
+
+			// Handle shared and non-shared rate limits differently
+			if isSharedRateLimit(route) {
+				addSharedRateLimitDescriptor(route, serviceDescriptors, domain, domainDescriptors)
+			} else {
+				addRouteSpecificDescriptor(route, serviceDescriptors, domain, domainDescriptors)
 			}
-			pbDescriptors = append(pbDescriptors, routeDescriptor)
 		}
 	}
 
-	if len(pbDescriptors) == 0 {
-		return nil
+	configs := createRateLimitConfigs(domainDescriptors)
+	return configs
+}
+
+// addSharedRateLimitDescriptor adds shared rate limit descriptors to the domain descriptor map
+func addSharedRateLimitDescriptor(
+	route *ir.HTTPRoute,
+	serviceDescriptors []*rlsconfv3.RateLimitDescriptor,
+	domain string,
+	domainDescriptors map[string][]*rlsconfv3.RateLimitDescriptor,
```

Comment:
is this domain created to avoid creating duplicates when parsing multiple Route IRs that contain the same shared bucket ? would it be enough for the map to hold the domain name without the actual descriptors, and skip parsing if that domain was already created ?